File: python/src/buble_ai/_streaming.py
from __future__ import annotations

import json
from typing import Any, AsyncIterable, AsyncIterator, Iterable, Iterator, List, Optional

from .types.chat import SSEEvent


def _parse_json(data: str) -> Any:
    if not data or data == "[DONE]":
        return None
    try:
        return json.loads(data)
    except json.JSONDecodeError:
        return None


def _event_from_lines(lines: List[str]) -> Optional[SSEEvent]:
    event_name = None
    data_lines: List[str] = []

    for raw in lines:
        line = raw.rstrip("\r")
        if not line or line.startswith(":"):
            continue
        if ":" in line:
            field, value = line.split(":", 1)
            value = value[1:] if value.startswith(" ") else value
        else:
            field, value = line, ""

        if field == "event":
            event_name = value
        elif field == "data":
            data_lines.append(value)

    if not event_name and not data_lines:
        return None

    data = "\n".join(data_lines)
    event: SSEEvent = {"data": data}
    if event_name:
        event["event"] = event_name
    parsed = _parse_json(data)
    if parsed is not None:
        event["json"] = parsed
    return event


def parse_sse_lines(lines: Iterable[str]) -> Iterator[SSEEvent]:
    block: List[str] = []
    for line in lines:
        if line == "":
            event = _event_from_lines(block)
            block = []
            if event:
                yield event
            continue
        block.append(line)

    event = _event_from_lines(block)
    if event:
        yield event


async def parse_async_sse_lines(lines: AsyncIterable[str]) -> AsyncIterator[SSEEvent]:
    block: List[str] = []
    async for line in lines:
        if line == "":
            event = _event_from_lines(block)
            block = []
            if event:
                yield event
            continue
        block.append(line)

    event = _event_from_lines(block)
    if event:
        yield event


def text_from_event(event: SSEEvent, protocol: str) -> Optional[str]:
    if event.get("data") == "[DONE]":
        return None
    payload = event.get("json")
    if not isinstance(payload, dict):
        return None

    if protocol == "openai":
        choices = payload.get("choices")
        if isinstance(choices, list) and choices:
            delta = choices[0].get("delta") if isinstance(choices[0], dict) else None
            if isinstance(delta, dict) and isinstance(delta.get("content"), str):
                return delta["content"]

    if protocol == "anthropic" and event.get("event") == "content_block_delta":
        delta = payload.get("delta")
        if isinstance(delta, dict) and isinstance(delta.get("text"), str):
            return delta["text"]

    if protocol == "gemini":
        chunks: List[str] = []
        candidates = payload.get("candidates")
        if isinstance(candidates, list) and candidates:
            content = candidates[0].get("content") if isinstance(candidates[0], dict) else None
            parts = content.get("parts") if isinstance(content, dict) else None
            if isinstance(parts, list):
                for part in parts:
                    if isinstance(part, dict) and isinstance(part.get("text"), str):
                        chunks.append(part["text"])
        return "".join(chunks) if chunks else None

    return None

File: rust/src/streaming.rs
use std::pin::Pin;

use bytes::Bytes;
use futures_core::Stream;
use futures_util::StreamExt;
use serde_json::Value;

use crate::error::{Error, Result};

/// Stream of parsed server-sent events.
pub type EventStream = Pin<Box<dyn Stream<Item = Result<SseEvent>> + Send>>;

/// Stream of extracted text chunks.
pub type TextStream = Pin<Box<dyn Stream<Item = Result<String>> + Send>>;

/// Protocol shape used to extract text from SSE payloads.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StreamProtocol {
    /// OpenAI-compatible chat completion chunks.
    OpenAi,
    /// Anthropic-compatible messages events.
    Anthropic,
    /// Gemini-compatible content chunks.
    Gemini,
}

/// Parsed server-sent event.
#[derive(Debug, Clone, PartialEq)]
pub struct SseEvent {
    /// Optional event name.
    pub event: Option<String>,
    /// Raw data payload.
    pub data: String,
    /// Parsed JSON payload, when `data` is JSON.
    pub json: Option<Value>,
}

pub(crate) fn events_from_bytes<S>(stream: S) -> EventStream
where
    S: Stream<Item = Result<Bytes>> + Send + 'static,
{
    Box::pin(async_stream::try_stream! {
        let mut stream = Box::pin(stream);
        let mut parser = SseParser::default();

        while let Some(chunk) = stream.next().await {
            let chunk = chunk?;
            let text = std::str::from_utf8(&chunk)
                .map_err(|error| Error::Stream(format!("invalid UTF-8 in SSE stream: {error}")))?;
            for event in parser.push(text)? {
                if event.data == "[DONE]" {
                    return;
                }
                yield event;
            }
        }

        for event in parser.finish()? {
            if event.data == "[DONE]" {
                return;
            }
            yield event;
        }
    })
}

#[derive(Debug, Default)]
struct SseParser {
    pending: String,
    event: Option<String>,
    data: Vec<String>,
}

impl SseParser {
    fn push(&mut self, chunk: &str) -> Result<Vec<SseEvent>> {
        self.pending.push_str(chunk);
        let mut output = Vec::new();

        while let Some(index) = self.pending.find('\n') {
            let mut line = self.pending[..index].to_string();
            self.pending.replace_range(..=index, "");
            if line.ends_with('\r') {
                line.pop();
            }
            if let Some(event) = self.push_line(&line)? {
                output.push(event);
            }
        }

        Ok(output)
    }

    fn finish(&mut self) -> Result<Vec<SseEvent>> {
        let mut output = Vec::new();
        if !self.pending.is_empty() {
            let line = std::mem::take(&mut self.pending);
            if let Some(event) = self.push_line(line.trim_end_matches('\r'))? {
                output.push(event);
            }
        }
        if let Some(event) = self.flush()? {
            output.push(event);
        }
        Ok(output)
    }

    fn push_line(&mut self, line: &str) -> Result<Option<SseEvent>> {
        if line.is_empty() {
            return self.flush();
        }
        if line.starts_with(':') {
            return Ok(None);
        }

        let (field, value) = line.split_once(':').map_or((line, ""), |(field, value)| {
            let value = value.strip_prefix(' ').unwrap_or(value);
            (field, value)
        });

        match field {
            "event" => self.event = Some(value.to_string()),
            "data" => self.data.push(value.to_string()),
            _ => {}
        }
        Ok(None)
    }

    fn flush(&mut self) -> Result<Option<SseEvent>> {
        if self.event.is_none() && self.data.is_empty() {
            return Ok(None);
        }

        let data = std::mem::take(&mut self.data).join("\n");
        let json = if data.is_empty() || data == "[DONE]" {
            None
        } else {
            Some(serde_json::from_str(&data)?)
        };
        Ok(Some(SseEvent {
            event: self.event.take(),
            data,
            json,
        }))
    }
}

pub(crate) fn text_from_event(event: &SseEvent, protocol: StreamProtocol) -> String {
    let Some(payload) = event.json.as_ref() else {
        return String::new();
    };

    match protocol {
        StreamProtocol::OpenAi => text_from_openai(payload),
        StreamProtocol::Anthropic => {
            if event.event.as_deref() != Some("content_block_delta") {
                return String::new();
            }
            text_from_anthropic(payload)
        }
        StreamProtocol::Gemini => text_from_gemini(payload),
    }
}

fn text_from_openai(payload: &Value) -> String {
    payload
        .get("choices")
        .and_then(Value::as_array)
        .and_then(|choices| choices.first())
        .and_then(|choice| choice.get("delta"))
        .and_then(|delta| delta.get("content"))
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_string()
}

fn text_from_anthropic(payload: &Value) -> String {
    payload
        .get("delta")
        .and_then(|delta| delta.get("text"))
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_string()
}

fn text_from_gemini(payload: &Value) -> String {
    let mut output = String::new();
    let Some(parts) = payload
        .get("candidates")
        .and_then(Value::as_array)
        .and_then(|candidates| candidates.first())
        .and_then(|candidate| candidate.get("content"))
        .and_then(|content| content.get("parts"))
        .and_then(Value::as_array)
    else {
        return output;
    };

    for part in parts {
        if let Some(text) = part.get("text").and_then(Value::as_str) {
            output.push_str(text);
        }
    }
    output
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_multiline_sse_event() {
        let mut parser = SseParser::default();
        let events = parser
            .push("event: content_block_delta\ndata: {\"delta\":{\"text\":\"Hi\"}}\n\n")
            .expect("parse event");
        assert_eq!(events.len(), 1);
        assert_eq!(events[0].event.as_deref(), Some("content_block_delta"));
        assert_eq!(text_from_event(&events[0], StreamProtocol::Anthropic), "Hi");
    }
}
